stdlib/public/core/KeyPath.swift

```diff
@@ -0,0 +1,1739 @@
+                 "not enough space for next component?")
+    let result = UnsafeRawBufferPointer(start: data.baseAddress, count: size)
+    data = UnsafeRawBufferPointer(
+      start: data.baseAddress.unsafelyUnwrapped + size,
+      count: data.count - size
+    )
+    return result
+  }
+}
+
+// MARK: Library intrinsics for projecting key paths.
+
+public // COMPILER_INTRINSIC
+func _projectKeyPathReadOnly<Root, Value>(
+  root: Root,
+  keyPath: KeyPath<Root, Value>
+) -> Value {
+  return keyPath.projectReadOnly(from: root)
+}
+
+public // COMPILER_INTRINSIC
+func _projectKeyPathWritable<Root, Value>(
+  root: UnsafeMutablePointer<Root>,
+  keyPath: WritableKeyPath<Root, Value>
+) -> (UnsafeMutablePointer<Value>, Builtin.NativeObject) {
+  return keyPath.projectMutableAddress(from: root)
+}
+
+public // COMPILER_INTRINSIC
+func _projectKeyPathReferenceWritable<Root, Value>(
+  root: Root,
+  keyPath: ReferenceWritableKeyPath<Root, Value>
+) -> (UnsafeMutablePointer<Value>, Builtin.NativeObject) {
+  return keyPath.projectMutableAddress(from: root)
+}
+
+// MARK: Appending type system
+
+// FIXME(ABI): The type relationships between KeyPath append operands are tricky
+// and don't interact well with our overriding rules. Hack things by injecting
+// a bunch of `appending` overloads as protocol extensions so they aren't
+// constrained by being overrides, and so that we can use exact-type constraints
+// on `Self` to prevent dynamically-typed methods from being inherited by
+// statically-typed key paths.
+public protocol _AppendKeyPath {}
+
+extension _AppendKeyPath where Self == AnyKeyPath {
+  public func appending(path: AnyKeyPath) -> AnyKeyPath? {
+    return _tryToAppendKeyPaths(root: self, leaf: path)
+  }
+}
+
+extension _AppendKeyPath /* where Self == PartialKeyPath<T> */ {
+  public func appending<Root>(path: AnyKeyPath) -> PartialKeyPath<Root>?
+  where Self == PartialKeyPath<Root> {
+    return _tryToAppendKeyPaths(root: self, leaf: path)
+  }
+  
+  public func appending<Root, AppendedRoot, AppendedValue>(
+    path: KeyPath<AppendedRoot, AppendedValue>
+  ) -> KeyPath<Root, AppendedValue>?
+  where Self == PartialKeyPath<Root> {
+    return _tryToAppendKeyPaths(root: self, leaf: path)
+  }
+  
+  public func appending<Root, AppendedRoot, AppendedValue>(
+    path: ReferenceWritableKeyPath<AppendedRoot, AppendedValue>
+  ) -> ReferenceWritableKeyPath<Root, AppendedValue>?
+  where Self == PartialKeyPath<Root> {
+    return _tryToAppendKeyPaths(root: self, leaf: path)
+  }
+}
+
+extension _AppendKeyPath /* where Self == KeyPath<T,U> */ {
+  public func appending<Root, Value, AppendedValue>(
+    path: KeyPath<Value, AppendedValue>
+  ) -> KeyPath<Root, AppendedValue>
+  where Self: KeyPath<Root, Value> {
+    return _appendingKeyPaths(root: self, leaf: path)
+  }
+
+  /* TODO
+  public func appending<Root, Value, Leaf>(
+    path: Leaf,
+    // FIXME: Satisfy "Value generic param not used in signature" constraint
```

Comment:
Looks like a bug in the 'all parameters are used' check -- the signature looks fine. Can you file a bug?